File: 63주차/표현가능한이진트리.js
function solution(numbers) {
  var answer = [];

  for (let v of numbers) {
    const bin = v.toString(2);
    const len = bin.length;
    const m = len.toString(2).length; // bin을 표현하기 위한 트리의 높이. 2*m -1 => 포화 이진 트리 일때 노드 개수.
    let s = "";
    s = "0".repeat(2 ** m - 1 - len) + bin; // 0을 앞에 채운다.

    if (!checkTree(s, 0, s.length - 1)) {
      // 부모가 0 인데, 자식이 1인경우 찾기.
      answer.push(0);
    } else answer.push(1);
  }
  return answer;
}

const checkTree = (tree, start, end) => {
  const mid = Math.floor((start + end) / 2); // 부모
  const left = Math.floor((start + mid - 1) / 2); // 왼쪽 자식
  const right = Math.floor((mid + 1 + end) / 2); // 오른쪽 자식.

  if (start === end) return true;

  if (tree[mid] === "0" && (tree[left] === "1" || tree[right] === "1")) {
    return false;
  }
  if (!checkTree(tree, start, mid - 1)) return false;
  if (!checkTree(tree, mid + 1, end)) return false;

  return true;
};

File: 68주차/표현가능한이진트리.py
import math
def check(num,start,end):
    root = (start+end)//2
    left = (start+(root-1))//2
    right = (root+1+end)//2
    if (start==end):# 부모가 0인데, 자식이 1이라면 표현할수 없는 이진트리.
        return True
    if (num[root] == '0'and( num[left]=='1'or num[right]=='1')):
        return False
    return check(num,start,root-1) and check(num,root+1,end)
    
def solution(numbers):
    answer = []
    for v in numbers:
        binNum = bin(v)[2:]
        h = int(math.log2(len(binNum)))+1
        totalLen = (2**h)-1
        binNum = '0'*(totalLen - len(binNum)) + binNum
        if check(binNum,0,totalLen-1):
            answer.append(1)
        else:
            answer.append(0)
    return answer